org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/CacheManager.java

```diff
@@ -129,55 +128,49 @@ public MavenTargetBundle getTargetBundle(Artifact artifact, BNDInstructions bndI
-
-			File folder = new File(baseDir, key);
-			return new CacheManager(folder);
-		});
+		return MANAGERS.computeIfAbsent(targetId, CacheManager::new);
 	}
 
 	/**
 	 * Clears all inactive cache locations that are older than the given time
 	 *
-	 * @param amount the amount of time to use
-	 * @param unit   the unit of time to use
+	 * @param timeout the duration since the last use after which a CacheManager is
+	 *                deleted
 	 */
-	public static synchronized void clearFilesOlderThan(long amount, TimeUnit unit) {
-		if (baseDir == null) {
-			throw new IllegalStateException("bundle not active");
-		}
-		File[] listFiles = baseDir.listFiles();
-		long deleteStamp = System.currentTimeMillis() - unit.toMillis(amount);
-		if (listFiles != null) {
-			for (File folder : listFiles) {
-				if (folder.isDirectory() && !MANAGERS.containsKey(folder.getName())) {
-					File marker = new File(folder, LASTACCESS_MARKER);
-					if (marker.isFile() && FileUtils.isFileOlder(marker, deleteStamp)) {
-						try {
-							FileUtils.deleteDirectory(folder);
-						} catch (IOException e) {
-						}
+	private static synchronized void clearFilesOlderThan(Path cacheBaseDir, Duration timeout) {
+		FileTime deleteBefore = FileTime.from(Instant.now().minus(timeout));
+		try (DirectoryStream<Path> folders = Files.newDirectoryStream(cacheBaseDir)) {
+			for (Path folder : folders) {
+				// Delete cache-folders last used before the specified timeout
+				// and not used in any manager
+				if (!MANAGERS.containsKey(folder.getFileName().toString())) {
+					Path marker = folder.resolve(LASTACCESS_MARKER);
+					if (Files.isRegularFile(marker) && Files.getLastModifiedTime(marker).compareTo(deleteBefore) < 0) {
+						Files.walkFileTree(folder, DeletingPathVisitor.withLongCounters());
 					}
 				}
 			}
+		} catch (IOException e) { // ignore exceptions
+			Platform.getLog(CacheManager.class).log(Status.error("Failed to clear Maven bundle cache", e));
 		}
 	}
 
-	/**
-	 * Set the basedir to use, purge all existing managers and invalidate them
-	 *
-	 * @param baseDir
-	 */
-	static synchronized void setBasedir(File baseDir) {
-		CacheManager.baseDir = baseDir;
-		for (CacheManager m : MANAGERS.values()) {
-			m.invalidated = true;
+	private static synchronized Path getCacheBaseDir() {
+		if (baseDir == null) {
+			Bundle bundle = FrameworkUtil.getBundle(CacheManager.class);
+			if (bundle == null) {
+				throw new IllegalStateException(CacheManager.class.getSimpleName() + " not loaded from a bundle");
+			}
+			baseDir = bundle.getDataFile("").toPath();
+			// clear all locations older than 14 days, this can be improved by
+			// 1) watch for changes in the workspace -> if target is deleted/removed from
+			// workspace we can clear the cache
+			// 2) we can add a preference page where the user can force clearing the cache
+			// or set the cache days
+			ForkJoinPool.commonPool().execute(() -> clearFilesOlderThan(baseDir, Duration.ofDays(14)));
```

Comment:
better use a dedicated thread here instead of the common pool

Reply:
What would be the benefit?

Reply:
Usually the common pool assumes that no blocking happens (e.g. wait for I/O) and your work is completely CPU bound, otherwise you might degrade the performance of other items using the common pool, see for example: 
https://www.jrebel.com/blog/parallel-java-streams

So for one-shot I/O performing it is generally better to just fire up a separate thread.

Reply:
Understand, makes sense. Thanks for the elaboration.